### PR TITLE
Follow symlink for --device argument.

### LINKED ISSUE #13840: Follow symlink for --device argument
A --device argument fails when the listed device is actually a symlink to a device.

Example:
`docker run --device /dev/disk/by-id/virtio-52525e7b-fe7b-4889-8:/dev/somevolume …etc…`
Fails with an error.
`Cannot start container 7edf77076ea5ea42b098728092c45bef24c9a742f3f122abf3b3bfd6bbaf2ef2: error gathering device information while adding custom device "/dev/disk/by-id/virtio-52525e7b-fe7b-4889-8": not a device node`

The `/dev/disk/by-id` entry is ofcourse a symlink to the actual device.
```$ ls -al /dev/disk/by-id/virtio-52525e7b-fe7b-4889-8
lrwxrwxrwx 1 root root 9 Jun 10 09:52 /dev/disk/by-id/virtio-52525e7b-fe7b-4889-8 -> ../../vdc```

The problem here is that I have no clue what the actual device path is when writing the docker run command in the service unit-file. It would be nice if docker could handle a device symlink, right now I have to manually resolve it to the correct device path.

There was an issue about this (#13706), but it was closed for a different reason.

uname -a
`Linux redactedhostname.novalocal 4.0.3 #2 SMP Thu May 14 00:34:07 UTC 2015 x86_64 Intel Xeon E312xx (Sandy Bridge) GenuineIntel GNU/Linux`

docker version
```Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2-dirty
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2-dirty
OS/Arch (server): linux/amd64```

docker -D info
```Containers: 1
Images: 9
Storage Driver: btrfs
 Build Version: Btrfs v3.17.1
 Library Version: 101
Execution Driver: native-0.2
Kernel Version: 4.0.3
Operating System: CoreOS 681.0.0
CPUs: 8
Total Memory: 15.68 GiB
Name: redactedhostname.novalocal
ID: S5RN:M3TJ:MSQ5:PMUC:QSET:5EMW:3FVJ:KHDR:RPBD:PHDG:JV3L:WCGV
Debug mode (server): false
Debug mode (client): true
Fds: 10
Goroutines: 19
System Time: Wed Jun 10 10:00:33 UTC 2015
EventsListeners: 0
Init SHA1: 75be3e87b0c63694ce2c69ab3f2c4251d6f69f83
Init Path: /usr/libexec/docker/dockerinit
Docker Root Dir: /var/lib/docker```

### PR DESCRIPTION
**Pull request by yongtang**
This pull request fixes issue #13840 where a --device argument fails when the listed device is actually a symlink to a device. It checks to see if the path is a symlink and, if it is, resolves the symlink and continue the operation with the resolved path.

The tests are done in this way. First a symlink is created to link to /dev/zero. Then this symlink is used to map devices in the container. Inside the container a buffer is created and the md5 is calculated:

``` sh
dd if=/dev/symzero bs=4K count=8 | md5sum
```

The expected md5 should be `bb7df04e1b0a2570657527a7e108ae23`.

Fixes: #13840

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Should we have tests for this?

**Comment by yongtang** [comment]
@thaJeztah Sure let me update the pull request.

**Comment by yongtang** [comment]
@thaJeztah Just added the tests. Please let me know if you see any other issues.

**Comment by timthelion** [comment]
Code looks good. It is certainly lucky that it was decided that `--device` and `--volume` should be separate. Otherwise, this PR would lead an absolutely devilish security problem:

Any volume that got mounted twice by the same image could contain a symlink to an otherwise forbidden device, because the first time the container was launched the symlink would be created, and the second time the container was launched, the symlink would be followed.

It might be worth adding a note in the docs that you can no longer safely use `--device` in an untrusted directory which might contain malicious symlinks...

**Comment by thaJeztah** [comment]
It's still in design review, so we could consider making this behavior optional, e.g. Add an extra option (`--device /foobar:follow-symlinks`).

**Comment by timthelion** [comment]
I don't think that there are many cases where the security flaw could happen, but the following one comes to mind, which I know exists in the wild:

````
$ docker run --volume=/dev/snd:/dev/snd --device=/dev/snd image-tag /path/to/executable
````

The reason why people do this is in order to get the `/dev/snd/by-path/` to show up in the container. The exploit would be to have the container create a symlink in `/dev/snd` which would point to some otherwise forbidden device, say `/dev/sda1`. The next time the container was run (or any container which was passed the `--device=/dev/snd` argument, for that matter) that new container would have access to `/dev/sda1`. Even though the user who issued `--volume=/dev/snd` never had the intention of giving containers access to `/dev/sda1`...

Perhaps there really should be a `:follow-symlinks` option, because we know that this can be exploited in reality, even though it is really the fault of the people doing `--volume=/dev/snd:/dev/snd` rather than `--volume=/dev/snd:/dev/snd:r`. And not Docker's fault at all...

**Comment by calavera** [comment]
LGTM

File: integration-cli/docker_cli_run_unix_test.go [hunk 1]
```diff
@@ -919,3 +919,39 @@ func (s *DockerSuite) TestRunSeccompWithDefaultProfile(c *check.C) {
...
+
+	defer os.RemoveAll(tmpDir)
+
+	// Create a symbolic link to /dev/zero
+	symZero := filepath.Join(tmpDir, "zero")
+	err = os.Symlink("/dev/zero", symZero)
+	c.Assert(err, checker.IsNil)
+
+	// Create a temporary file, then a symbolic link to the temporary file
+	tmpFile := filepath.Join(tmpDir, "zero")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_unix_test.go]: Can you explain this?
`tmpFile` is the same as `symZero` and then we're writing to it. Can we skip the `tmpFile` parts and instead use `symZero`?

Comment (yongtang) [inline integration-cli/docker_cli_run_unix_test.go]: @cpuguy83:
Sorry my bad. That was a bug/typo.
There are two tests:
In the first test is to create a symlink (tmpDir/zero) pointing to /dev/zero, then inside the container do a md5 calculation. This test is to show symlink to a device works.
In the second test, I was meant to create a temporary file, write some data into this temporary file, then create a symlink (tmpDir/file) pointing to the temporary file created. This test was meant to show that symlink to a file (not a device) will not work. (The name of the tmpFile should be something else, not filepath.Join(tmpDir, "zero"))
I will update the pull request shortly.

**Comment by yongtang** [comment]
@cpuguy83 Just updated the pull request. Let me know there are any other issues.

**Pull request merged**